### PR TITLE
Fixes width/height extraction with respect to exif orientation tag

### PR DESCRIPTION
**Pull request by alexxxwork**
### Details

Expensify uses react-native-fast-image and added a patch to be able to get dimensions of the image displayed in the component on an onLoad event. But this method doesn't work correctly with images that have an exif property of image orientation. In such cases image is displayed properly (rotated according to the exif orientation flag) but dimensions returned from react-native-fast-image component are wrong (width/height swapped).  
To resolve this propblem we should add an exif reader to get orientation tag and correct dimensions in such cases.

### Fixed Issues

$ https://github.com/Expensify/App/issues/14448
$ https://github.com/Expensify/App/issues/14448#issuecomment-1422330478

### Tests

1. Launch the app
2. Login with any account
3. Go to any chat
4. Tap on plus button
5. Select Add Attachment
6. Select Take a photo
7. Take a photo and send it

- [x] Verify that no errors appear in the JS console
- [x] Picture should be displayed in full size
- [x]  upload image from iOS camera and preview in android
- [x] upload image from iOS gallery and preview in android
- [x] upload image from web and preview in android
- [x] upload image from android camera and preview in all platforms
- [x] upload image from android gallery and preview in all platforms

### Offline tests

1. Go to a chat with previously uploaded pictures
2.  Pictures should be displayed in full size

### QA Steps

1. Launch the app
2. Login with any account
3. Go to any chat
4. Tap on plus button
5. Select Add Attachment
7. Select Take a photo
8. Take a photo and send it

- [x] Verify that no errors appear in the JS console
- [x] Picture should be displayed in full size
- [x]  upload image from iOS camera and preview in android
- [x] upload image from iOS gallery and preview in android
- [x] upload image from web and preview in android
- [x] upload image from android camera and preview in all platforms
- [x] upload image from android gallery and preview in all platforms

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

image

</details>

<details>
<summary>Mobile Web - Chrome</summary>

Screenshot_20230210_030608_Chrome

</details>

<details>
<summary>Mobile Web - Safari</summary>

screen_web_ios

</details>

<details>
<summary>Desktop</summary>

image

</details>

<details>
<summary>iOS</summary>

screen_ios

</details>

<details>
<summary>Android</summary>

Screenshot_20230210_040127_New Expensify

</details>

### CODE REVIEW

**Comment by alexxxwork** [comment]
I have read the CLA Document and I hereby sign the CLA

**Comment by alexxxwork** [comment]
> @alexxxwork please let us know when PR is ready for review. Also, please make sure to follow PR template exactly from next time. Reviewers are not linked correctly. pullerbear should have correctly assigned @mountiny and me as reviewers.

Seems like a don't have enough rights to request review from specific persons. Should I just mention desired reviewers @mountiny @0xmiroslav? The PR is ready for review

**Comment by alexxxwork** [comment]
> @alexxxwork I believe you didn't comment like this. Can you share your screenshot on editing state? template

Do you mean that mention in comment above details header?

```
<!-- If necessary, assign reviewers that know the area or changes well. Feel free to tag any additional reviewers you see fit. @mountiny @0xmiroslav  -->
### Details

### Fixed Issues
<!---
1. Please replace GH_LINK with a URL link to the GitHub issue this Pull Request is fixing.
2. Please replace PROPOSAL: GH_LINK_ISSUE(COMMENT) with a URL link to your GitHub comment, which contains the approved proposal (i.e. the proposal that was approved by Expensify).

Do NOT add the special GH keywords like `fixed` etc, we have our own process of managing the flow.
It MUST be an entire link to the github issue and your comment proposal ; otherwise, the linking will not work as expected.

Make sure this section looks similar to this (you can link multiple issues using the same formatting, just add a new line):

$ https://github.com/Expensify/App/issues/<number-of-the-issue>
$ https://github.com/Expensify/App/issues/<number-of-the-issue(comment)>

Do NOT only link the issue number like this: $ #<number-of-the-issue>
--->
$ [#14448](https://github.com/Expensify/App/issues/14448)   
PROPOSAL: [#14448 (comment)](https://github.com/Expensify/App/issues/14448#issuecomment-1422330478)
```

**Comment by alexxxwork** [comment]
> @alexxxwork ok so that's the issue. Automation didn't work because you didn't follow template.
> 
> ```diff
> - $ [#14448](https://github.com/Expensify/App/issues/14448) 
> + $ https://github.com/Expensify/App/issues/14448
> ```
> 
> I understand this is your first PR but from next job, please follow template correctly.

:facepalm: Thought that this automation worked correctly as long as the link to PR appeared in the issue.

**Comment by alexxxwork** [comment]
> @alexxxwork ok so that's the issue. Automation didn't work because you didn't follow template.
> 
> ```diff
> - $ [#14448](https://github.com/Expensify/App/issues/14448) 
> + $ https://github.com/Expensify/App/issues/14448
> ```
> 
> I understand this is your first PR but from next job, please follow template correctly.

Should I recreate this PR to trigger correct assignment?

**Review by 0xmiroslav — CHANGES_REQUESTED** [review]

File: patches/react-native-fast-image+8.6.3.patch [hunk 1]
```diff
@@ -32,13 +33,21 @@ index 0000000..03ad017
 +        BitmapFactory.Options bitmapOptions = new BitmapFactory.Options();
 +        bitmapOptions.inJustDecodeBounds = true;
 +        BitmapFactory.decodeStream(source, null, bitmapOptions);
++        source.reset();
```
[the remark this answers is not in the document]
Comment (alexxxwork) [reply patches/react-native-fast-image+8.6.3.patch:36]: This function leaves stream's position where ever it was after reading the encoded data (https://developer.android.com/reference/android/graphics/BitmapFactory#decodeStream(java.io.InputStream)) so we need to rewind the stream to be able to read image header with exif values

**Comment by mountiny** [comment]
@alexxxwork no problem as @0xmiroslav mentioned the issue needs to be linked exactly as the template says to make this work.

Can you also fill out the details section with some summary of what are we doing here?

Also will we make an upstream PR?

**Comment by mountiny** [comment]
@0xmiroslav oh sorry, thanks for the link! Thank you 🙇

**Comment by alexxxwork** [comment]
> Can you also fill out the details section with some summary of what are we doing here?

@mountiny updated details

File: patches/react-native-fast-image+8.6.3.patch [hunk 2]
```diff
@@ -257,15 +268,15 @@ diff --git a/node_modules/react-native-fast-image/ios/FastImage/FFFastImageView.
 index f710081..391ef92 100644
 --- a/node_modules/react-native-fast-image/ios/FastImage/FFFastImageView.m
 +++ b/node_modules/react-native-fast-image/ios/FastImage/FFFastImageView.m
-@@ -54,7 +54,6 @@ - (void) setOnFastImageError: (RCTDirectEventBlock)onFastImageError {
```
[the remark this answers is not in the document]
Comment (alexxxwork) [reply patches/react-native-fast-image+8.6.3.patch:260]: haven't toched them. Could be unix vs windows

[the remark this answers is not in the document]
Comment (alexxxwork) [reply patches/react-native-fast-image+8.6.3.patch:260]: I've regenerated patch with npx patch-package

**Comment by alexxxwork** [comment]
> @alexxxwork please replace with this file and commit again [react-native-fast-image+8.6.3.patch](https://github.com/Expensify/App/files/10707196/react-native-fast-image%2B8.6.3.patch)

Seems to be ok now

**Review by mountiny — COMMENTED** [review]
Looking good so far!

**Review by 0xmiroslav — APPROVED** [review]
windows

https://user-images.githubusercontent.com/97473779/218221352-d8955125-9796-47d7-9cfe-612027804960.mov


https://user-images.githubusercontent.com/97473779/218221360-4e9ae8e3-4684-41f5-8462-cc08efe6c16f.mov

tested various cases: windows/mac images rotation, android/iOS camera/gallery portrait/landscape images and all look good.
I attached before-fix and after-fix videos in Android Screenshots. All other platforms look same between before-fix and after-fix.

**Review by mountiny — APPROVED** [review]
@0xmiroslav I do appreciate your thorough testing here!

@alexxxwork Congratulations, you first PR done, onwards and upwards 😍

**Pull request merged**